### PR TITLE
[frame/im-online] remove network state from heartbeats

### PR DESCRIPTION
**Pull request by melekes**
Users should use DHT for discovering new nodes. The reason for adding external addresses was unstable work of authority discovery (see https://github.com/paritytech/substrate/issues/2719), which is now stable. Hence we can safely remove `external_addresses`.

Refs https://github.com/paritytech/polkadot-sdk/issues/646
polkadot companion: https://github.com/paritytech/polkadot/pull/7309

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2908457

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2908564

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2908730

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2908831

**Comment by dmitry-markin** [comment]
They are not used for discovery, but are we sure that they are not used for something else?

**Review by dmitry-markin — APPROVED** [review]
LGTM code-wise, but I don't know if external addresses in heartbeats are used for anything, and it's safe to remove them.

**Comment by melekes** [comment]
> They are not used for discovery, but are we sure that they are not used for something else?

I'm going to refer to this comment: https://github.com/paritytech/polkadot/issues/7181#issuecomment-1536144129. Not merging this in the closest release is also fine with me (to give developers some time to upgrade, but we should probably notify them somehow...)

**Review by altonen — COMMENTED** [review]
Looks good to me as well but I don't know if these are used for anything else so I'd wait for someone else's approval

File: primitives/core/src/offchain/mod.rs
Comment (bkchr) [inline primitives/core/src/offchain/mod.rs:193]: This would break the host function. Please revert.

File: client/offchain/src/api.rs
Comment (bkchr) [inline client/offchain/src/api.rs:19]: All the changes in here should also be reverted. The point was about removing the stuff from `im-online` and not everywhere else.

File: frame/im-online/src/lib.rs
Comment (bkchr) [inline frame/im-online/src/lib.rs:452]: We should get rid of the entire network_state. In the heartbeat and everywhere in this pallet.

**Review by bkchr — CHANGES_REQUESTED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2916191

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2916296

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2916978

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2916977

**Comment by burdges** [comment]
We do want this information but this was never a good way to get it, so afaik this going away makes sense.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2920226

**Review by bkchr — APPROVED** [review]

**Comment by shawntabrizi** [comment]
Is there a missing storage migration?

**Comment by bkchr** [comment]
> Is there a missing storage migration?

No, we never read `ReceivedHeartbeats` we only call `contains_key` and that doesn't decode the value.

**Comment by bkchr** [comment]
bot merge

File: frame/im-online/src/lib.rs
Comment (ggwpez) [inline frame/im-online/src/lib.rs:347]:
```suggestion
		StorageDoubleMap<_, Twox64Concat, SessionIndex, Twox64Concat, AuthIndex, ()>;
```
Why a bool when only `true` is inserted?

Comment (kianenigma) [reply frame/im-online/src/lib.rs:347]: This ^^

**Comment by shawntabrizi** [comment]
> > Is there a missing storage migration?
> 
> No, we never read `ReceivedHeartbeats` we only call `contains_key` and that doesn't decode the value.

Yes. But it looks like data is only cleaned up at the end of sessions?

There will be one session, where data in the storage map will be of two different types. Seems to be no issue to the runtime, but might be an issue to block explorers and other UIs.

Just a consideration, obv not the biggest deal.

File: frame/im-online/src/lib.rs
Comment (bkchr) [reply frame/im-online/src/lib.rs:347]: Do we even support this? @cheme could we store empty values in the trie?

Comment (melekes) [reply frame/im-online/src/lib.rs]: Thought so too. Changed in https://github.com/paritytech/substrate/pull/14251/commits/dcc7b38fa3e53876a0e1d5c76b6c3ef64c157c76

Comment (bkchr) [reply frame/im-online/src/lib.rs]: Did you ensured this works @melekes?

**Comment by melekes** [comment]
> There will be one session, where data in the storage map will be of two different types. Seems to be no issue to the runtime, but might be an issue to block explorers and other UIs.

@bkchr do you want me to write a storage migration here?

**Comment by melekes** [comment]
> > > There will be one session, where data in the storage map will be of two different types. Seems to be no issue to the runtime, but might be an issue to block explorers and other UIs.
> > 
> > 
> > @bkchr do you want me to write a storage migration here?
> 
> I personally think it is fine without a migration. However, @shawntabrizi is probably right and indexers would be happy. So, yes please write a small one.

is this https://docs.substrate.io/reference/how-to-guides/storage-migrations/basic-storage-migration/ the actual guide to writing one?

**Review by sam0x17 — APPROVED** [review]

**Comment by bkchr** [comment]
You can take parts of the guide, but you should implement the Onruntimeupgrade trait on your own for a "free standing" migration type. Just check the migrations of other pallets.

File: frame/im-online/src/lib.rs
Comment (melekes) [reply frame/im-online/src/lib.rs]: I thought (maybe naively) that 500 lines of tests should cover that, no?

File: frame/im-online/src/migration.rs
Comment (melekes) [inline frame/im-online/src/migration.rs:74]: if this `Config` param is now gone, how do I configure `BoundedOpaqueNetworkState`? Should I copy original `Config` here and use it instead?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2949582

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2949654

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2950814

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2951261

File: frame/im-online/src/lib.rs
Comment (melekes) [reply frame/im-online/src/lib.rs]: Actually it does not work. I'm going to revert https://github.com/paritytech/substrate/commit/dcc7b38fa3e53876a0e1d5c76b6c3ef64c157c76

**Comment by melekes** [comment]
could @bkchr or @paritytech/frame-coders please take a look at the migration https://github.com/paritytech/substrate/pull/14251/files#diff-e8faeeb1f7c80155f21fe0f315f67a539ce51f8c08bed0d90130aae3cca6627c before we merge this?

File: frame/im-online/src/migration.rs
Comment (melekes) [inline frame/im-online/src/migration.rs:89]: I've noticed there's a difference between iterating over keys and values. If so, when should I iterate over values? over keys?

File: frame/im-online/src/migration.rs
Comment (bkchr) [inline frame/im-online/src/migration.rs:58]:
```suggestion
		/// PeerId of the local node in SCALE encoded.
		pub peer_id: Vec<u8>,
		/// List of addresses the node knows it can be reached as.
		pub external_addresses:
			Vec<Vec<u8>>,
```

No need to use `WeakBoundedVec` and makes the code easier.

File: frame/im-online/src/migration.rs
Comment (bkchr) [inline frame/im-online/src/migration.rs:35]: These types can just be moved to v1.

File: frame/im-online/src/migration.rs
Comment (bkchr) [inline frame/im-online/src/migration.rs:105]: This is bad, because you iterate here once over the state and then below again. You should merge this.

Comment (bkchr) [reply frame/im-online/src/migration.rs:105]: `DoubleMap` also directly provides a `translate` method.

File: frame/im-online/src/migration.rs
Comment (melekes) [reply frame/im-online/src/migration.rs:35]: I like that `v0` emphasises previous version, which no longer exists. Also I copied it from another frame pallet -> better to stay consistent I think

**Comment by melekes** [comment]
bot merge

File: frame/im-online/src/migration.rs
Comment (bkchr) [inline frame/im-online/src/migration.rs:114]: This should have been `>=`. To make it possible that CI checks not directly start to fail when v1 and a potential v2 in the future would be part of runtime.

File: frame/im-online/src/migration.rs
Comment (melekes) [reply frame/im-online/src/migration.rs:122]: what about try-runtime CLI cmd? Or it's not used? I've seen at least 15 migration.rs files containing that flag.

File: frame/im-online/src/migration.rs
Comment (bkchr) [reply frame/im-online/src/migration.rs:114]: Ahh, I also have overseen this :see_no_evil: 

So, the general idea is that these migrations may stay in the runtime for quite some time and we also have tests on Polkadot that run these migrations using try runtime. Now, when the migration would be applied (which could be at any time) tests will start to fail. TLDR: The check should be removed from `pre_upgrade` and changed in post as said above.

**Pull request merged**
